examples/infix_math_parser.py

```diff
@@ -0,0 +1,190 @@
+"""Defines a recursive parser for parsing mathematical expressions in infix notation.
+
+Supports binary, unary, and variadic operations. These can also be customized
+in the InfixExpressionParser class variables. Utilizes some regex to improve its
+performance.
+
+Examples of parsing:
+
+The expression "f_1 + f_2 - 1e-3" is parsed into [['f_1', '+', 'f_2', '-', 0.001]]
+
+The expression "Max(Ln(x) + Lb(Abs(y)), Ceil(Sqrt(garlic) * 3), (potato ** 2) / 4, Abs(cosmic) + 10)"
+is parsed into
+[['Max', [[['Ln', ['x']], '+', ['Lb', [['Abs', ['y']]]]], ['Ceil', [[['Sqrt', ['garlic']], '*', 3]]], [['potato', '**', 2], '/', 4], [['Abs', ['cosmic']], '+', 10]]]]
+"""
+
+from operator import or_
+from typing import ClassVar
+
+from pyparsing import (
+    Forward,
+    Group,
+    Literal,
+    ParserElement,
+    Suppress,
+    delimitedList,
+    infixNotation,
+    one_of,
+    opAssoc,
+    pyparsing_common,
+    Regex,
+)
+
+# Enable Packrat for better performance in recursive parsing
+ParserElement.enablePackrat(None)
+
+
+class InfixExpressionParser:
+    """A class for defining an infix notation parsers."""
+
+    # Supported infix binary operators, i.e., '1+1'. The key is the notation of the operator in infix format,
+    # and the value the notation in parsed format.
+    BINARY_OPERATORS: ClassVar[dict] = {
+        "+": "Add",
+        "-": "Subtract",
+        "*": "Multiply",
+        "/": "Divide",
+        "**": "Power",
+    }
+
+    # Supported infix unary operators, i.e., 'Cos(90)'. The key is the notation of the operator in infix format,
+    # and the value the notation in parsed format.
+    UNARY_OPERATORS: ClassVar[dict] = {
+        "Cos": "Cos",
+        "Sin": "Sin",
+        "Tan": "Tan",
+        "Exp": "Exp",
+        "Ln": "Ln",
+        "Lb": "Lb",
+        "Lg": "Lg",
+        "LogOnePlus": "LogOnePlus",
+        "Sqrt": "Sqrt",
+        "Square": "Square",
+        "Abs": "Abs",
+        "Ceil": "Ceil",
+        "Floor": "Floor",
+        "Arccos": "Arccos",
+        "Arccosh": "Arccosh",
+        "Arcsin": "Arcsin",
+        "Arcsinh": "Arcsinh",
+        "Arctan": "Arctan",
+        "Arctanh": "Arctanh",
+        "Cosh": "Cosh",
+        "Sinh": "Sinh",
+        "Tanh": "Tanh",
+        "Rational": "Rational",
+        "-": "Negate",
```

Comment:
I don't think this will ever evaluate as "Negate", since "-" is a signop.

Reply:
This is true. This is an artifact from the application I have worked and I extracted this example from.